csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
   override ForStmt stmt;
 
   override TranslatedElement getChild(int id) {
-    id = 0 and result = this.getDeclAndInit()
+    initializerIndices(id) and result = this.getDeclAndInit(id)
     or
-    id = 1 and result = this.getCondition()
+    updateIndices(id) and result = this.getUpdate(id - initializersNo())
     or
-    id = 2 and result = this.getUpdate()
+    id = initializersNo() + updatesNo() and result = this.getCondition()
     or
-    id = 3 and result = this.getBody()
+    id = initializersNo() + updatesNo() + 1 and result = this.getBody()
   }
 
-  private TranslatedLocalDeclaration getDeclAndInit() {
-    result = getTranslatedLocalDeclaration(stmt.getAnInitializer())
+  private TranslatedLocalDeclaration getDeclAndInit(int index) {
+    result = getTranslatedLocalDeclaration(stmt.getInitializer(index))
   }
 
   private predicate hasInitialization() { exists(stmt.getAnInitializer()) }
 
-  TranslatedExpr getUpdate() { result = getTranslatedExpr(stmt.getAnUpdate()) }
+  TranslatedExpr getUpdate(int index) { result = getTranslatedExpr(stmt.getUpdate(index)) }
 
   private predicate hasUpdate() { exists(stmt.getAnUpdate()) }
 
+  private int initializersNo() { result = count(stmt.getAnInitializer()) }
+
+  private int updatesNo() { result = count(stmt.getAnUpdate()) }
+
+  private predicate initializerIndices(int index) { index in [0 .. initializersNo() - 1] }
+
+  private predicate updateIndices(int index) {
+    index in [initializersNo() .. initializersNo() + updatesNo() - 1]
+  }
```

Comment:
I think I would rewrite to
```
  private int updateIndex(int index) {
    result in [0 .. updatesNo() - 1] and
    index = initializersNo() + result
  }
```
and then change
```
updateIndices(id) and result = this.getUpdate(id - initializersNo())
```
above to
```
result = this.getUpdate(updateIndex(id))
```

Reply:
That seems clearer, refactored that bit.